_build/lexicon/checklexicon.php

```diff
@@ -0,0 +1,565 @@
+        }
+
+        if (file_exists($this->setupLexiconPath . $this->language . '/')) {
+            $iterator = new \DirectoryIterator($this->setupLexiconPath . $this->language . '/');
+            foreach ($iterator as $path => $current) {
+                if (strpos($current->getFilename(), 'inc.php') !== false) {
+                    try {
+                        include $current->getRealPath();
+                    } catch (Exception $e) {
+                        $this->invalidLexicons[] = $current->getFilename();
+                    }
+                }
+            }
+        } else {
+            return 'Could not load the lexicons in the setup language folder "' . $this->setupLexiconPath . $this->language . '/' . '"!';
+        }
+
+        return $_lang;
+    }
+
+    /**
+     * Add used lexicon keys
+     */
+    private function addKeys()
+    {
+        $directory = new \RecursiveDirectoryIterator($this->scanPath, \RecursiveDirectoryIterator::SKIP_DOTS);
+        $filter = new \RecursiveCallbackFilterIterator($directory, function ($current, $key, $iterator) {
+            /** @var \RecursiveDirectoryIterator $current */
+            if ($current->getFilename()[0] === '.') {
+                return false;
+            }
+            if ($current->isDir()) {
+                return !in_array($current->getFilename(), $this->excludedFolders);
+            } else {
+                $pathinfo = pathinfo($current->getFilename());
+                return ($current->isFile() && isset($pathinfo['extension']) && (
+                        $pathinfo['extension'] == 'php' ||
+                        $pathinfo['extension'] == 'js' ||
+                        $pathinfo['extension'] == 'html' ||
+                        $pathinfo['extension'] == 'tpl' ||
+                        $pathinfo['basename'] == 'config.json'
+                    ) &&
+                    strpos($pathinfo['basename'], 'min.js') === false &&
+                    strpos($pathinfo['basename'], 'ext-') !== 0
+                ) ? true : false;
+            }
+        });
+        $iterator = new \RecursiveIteratorIterator($filter);
+
+        foreach ($iterator as $path => $current) {
+            $this->addPhpKeys($path);
+            $this->addJsKeys($path);
+            $this->addChunkKeys($path);
+            $this->addSmartyKeys($path);
+        }
+        $this->addSettingKeys();
+        $this->addMenuKeys();
+        $this->addWidgetKeys();
+        $this->addPermissionKeys();
+
+        $this->languageKeys = array_unique($this->languageKeys);
+        sort($this->languageKeys);
+    }
+
+    /**
+     * Add lexicon calls in php files:
+     * modx->lexicon('whatever'
+     *
+     * @param string $filename
+     */
+    private function addPhpKeys($filename)
+    {
+        $fileContent = file_get_contents($filename);
+        $results = array();
+        preg_match_all('/(modx|xpdo)->lexicon\((["\'])(.*?)\2\s*[,\)]/m', $fileContent, $results);
```

Comment:
We should consider using [Named Capturing Groups](https://www.regular-expressions.info/named.html) here.